File: typescript/src/model/types/wdl-pair-type.ts
/** Pair type nodes. */
import { WdlType, WdlTypeComponentType } from './wdl-type.js';

/** Models a pair type such as `Pair[String, Int]`. */
export class WdlPairType extends WdlType {
  /** Creates a pair type from its left type, right type, and optional flag. */
  public constructor(
    private pairLeftType?: WdlType,
    private pairRightType?: WdlType,
    optional = false,
  ) {
    super(optional);
  }

  /** Returns the broad type family for this node. */
  public componentType(): WdlTypeComponentType {
    return WdlTypeComponentType.PAIR;
  }

  /** Returns the left member type. */
  public leftType(): WdlType | undefined {
    return this.pairLeftType;
  }

  /** Sets the left member type. */
  public setLeftType(pairLeftType: WdlType | undefined): void {
    this.pairLeftType = pairLeftType;
  }

  /** Returns the right member type. */
  public rightType(): WdlType | undefined {
    return this.pairRightType;
  }

  /** Sets the right member type. */
  public setRightType(pairRightType: WdlType | undefined): void {
    this.pairRightType = pairRightType;
  }
}

File: java/src/main/java/com/myriad/wdl/model/types/WdlPairType.java
package com.myriad.wdl.model.types;

/** Pair[L, R] type node. */
public final class WdlPairType extends WdlType {
  private WdlType pairLeftType;
  private WdlType pairRightType;

  public WdlPairType() {
    super();
  }

  public WdlPairType(WdlType pairLeftType, WdlType pairRightType, boolean optional) {
    super(optional);
    this.pairLeftType = pairLeftType;
    this.pairRightType = pairRightType;
  }

  @Override
  public ComponentType componentType() {
    return ComponentType.PAIR;
  }

  /** Returns the left pair member type. */
  public WdlType leftType() {
    return pairLeftType;
  }

  /** Sets the left pair member type. */
  public void setLeftType(WdlType pairLeftType) {
    this.pairLeftType = pairLeftType;
  }

  /** Returns the right pair member type. */
  public WdlType rightType() {
    return pairRightType;
  }

  /** Sets the right pair member type. */
  public void setRightType(WdlType pairRightType) {
    this.pairRightType = pairRightType;
  }
}
